File: tests/fixtures/ruby-simple/lib/payments/payment_service.rb
# Handles payment processing with logging and validation.
require_relative "../utils/logger"
require_relative "payment_interface"

module Payments
  class PaymentService
    include PaymentInterface

    attr_reader :logger
    attr_accessor :retry_count

    def initialize(logger)
      @logger = logger
      @retry_count = 3
    end

    def process_payment(order_id, amount)
      logger.info("Processing payment for order: #{order_id}")
      if amount <= 0
        logger.error("Invalid amount: #{amount}")
        return false
      end
      execute_transaction(order_id, amount)
    end

    def refund_payment(transaction_id)
      logger.info("Refunding transaction: #{transaction_id}")
      true
    end

    def self.from_config(config)
      new(Utils::Logger.new(config[:context]))
    end

    private

    def execute_transaction(order_id, amount)
      logger.info("Executing transaction for #{order_id}")
      true
    end
  end
end

File: tests/fixtures/php-simple/src/Payments/PaymentService.php
<?php

namespace App\Payments;

use App\Utils\Logger;

/** Handles payment processing with logging and validation. */
class PaymentService implements PaymentServiceInterface
{
    private Logger $logger;
    protected int $retryCount = 3;

    public function __construct(Logger $logger)
    {
        $this->logger = $logger;
    }

    public function processPayment(string $orderId, float $amount): bool
    {
        $this->logger->info("Processing payment for order: {$orderId}");
        if ($amount <= 0) {
            $this->logger->error("Invalid amount: {$amount}");
            return false;
        }
        return $this->executeTransaction($orderId, $amount);
    }

    public function refundPayment(string $transactionId): void
    {
        $this->logger->info("Refunding transaction: {$transactionId}");
    }

    public static function createDefault(): self
    {
        return new self(new Logger("PaymentService"));
    }

    private function executeTransaction(string $orderId, float $amount): bool
    {
        $this->logger->info("Executing transaction for {$orderId}");
        return true;
    }
}
